pkg/lib/collections/priority_queue.go

```diff
@@ -0,0 +1,231 @@
+	}
+	heap.Init(&q.internalQueue)
+	return q
+}
+
+// Enqueue will add the item specified by `data` to the queue with the
+// the priority given by `priority`.
+func (pq *PriorityQueue[T]) Enqueue(data T, priority int) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+
+	pq.enqueue(data, priority)
+}
+
+// enqueue is a lock-free version of Enqueue for internal use when a
+// method already has a lock.
+func (pq *PriorityQueue[T]) enqueue(data T, priority int) {
+	heap.Push(
+		&pq.internalQueue,
+		&heapItem{
+			value:    data,
+			priority: priority,
+		},
+	)
+}
+
+// Dequeue returns the next highest priority item, returning both
+// the data Enqueued previously, and the priority with which it was
+// enqueued. An err (ErrEmptyQueue) may be returned if the queue is
+// currently empty.
+func (pq *PriorityQueue[T]) Dequeue() *QueueItem[T] {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+
+	return pq.dequeue()
+}
+
+// dequeue is a lock-free version of Dequeue() for use by internal
+// methods that already have a lock
+func (pq *PriorityQueue[T]) dequeue() *QueueItem[T] {
+	if pq.IsEmpty() {
+		return nil
+	}
+
+	internalItem := heap.Pop(&pq.internalQueue)
+	heapItem := internalItem.(*heapItem)
+	item, _ := heapItem.value.(T)
+
+	return &QueueItem[T]{Value: item, Priority: heapItem.priority}
+}
+
+// DequeueWhere allows the caller to iterate through the queue, in priority order, and
+// attempt to match an item using the provided `MatchingFunction`.  This method has a high
+// time cost as dequeued but non-matching items must be held and requeued once the process
+// is complete.  Luckily, we use the same amount of space (bar a few bytes for the
+// extra PriorityQueue) for the dequeued items.
+func (pq *PriorityQueue[T]) DequeueWhere(matcher MatchingFunction[T]) *QueueItem[T] {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+
+	var result *QueueItem[T] = nil
+
+	// Create a new array to hold items that are not matches, this is suboptimal for time
+	// but not really an issue for space as the items we add here will have been removed
+	// from the queue.
+	unmatched := make([]*QueueItem[T], 0, pq.Len())
+
+	// Keep dequeueing items until one of them matches the function provided.
+	// If any match it will be returned after the other items have been requeued.
+	// If any iteration does not generate a match, the item is requeued in a temporary
+	// queue reading for requeueing on this queue later on.
+	for pq.internalQueue.Len() > 0 {
+		qitem := pq.dequeue()
+
+		if qitem == nil {
+			return nil
+		}
+
+		if matcher(qitem.Value) {
+			result = qitem
+			break
+		}
+
+		// Add to the queue
+		unmatched = append(unmatched, qitem)
+	}
+
+	// Re-add the items that were not matched back onto the Q
+	lo.ForEach(unmatched, func(item *QueueItem[T], _ int) {
+		pq.enqueue(item.Value, item.Priority)
+	})
+
+	// return the result we found, which might still be nil (not found)
+	return result
+}
+
+func (pq *PriorityQueue[T]) Merge(other *PriorityQueue[T]) {
```

Comment:
not used at the moment, but okay to leave it if you think useful

Reply:
Meant to remove it actually, trivial enough to re-add if needed later.